### ISSUE TITLE
Consensus (bank hash) mismatch between v1.13/v1.14 and v1.15/master (mnb)

### ISSUE THREAD

**Issue by steviez** [solana-labs/solana#30375]
#### Problem

Nodes running v1.15.x and/or the tip of master were forked off from mainnet; v1.13 and v1.14 were consistent with each other. That is, the change in behavior is limited to v1.15/master branches.
```
Cluster duplicate confirmed slot 178114337 with hash GkTuscdwprNteWUMUjNRCJxZf3Gg55bt1tzV8QkhKgv5,
but our version has hash 3ncx1EbLUoGPQUoNgwpqYJ8cmaK3BiUwwzN7tLJJ5rmQ
```
Nodes then panicked after repeatedly reproducing the incorrect hash:
```
thread 'solReplayStage' panicked at 'We have tried to repair duplicate slot: 178114337 more than 10 times
and are unable to freeze a block with bankhash GkTuscdwprNteWUMUjNRCJxZf3Gg55bt1tzV8QkhKgv5,
instead we have a block with bankhash Some(3ncx1EbLUoGPQUoNgwpqYJ8cmaK3BiUwwzN7tLJJ5rmQ).
This is most likely a bug in the runtime. At this point manual intervention is needed to make progress. Exiting',
```

The correct bank hash from a `v1.14` node:
```
bank frozen: 178114337
hash: GkTuscdwprNteWUMUjNRCJxZf3Gg55bt1tzV8QkhKgv5
accounts_delta: BiDNt3K62GRu1yvfkkAj1cXULZk6T3E1mb3sXq3Kv7oV
signature_count: 2261
last_blockhash: CbE1VP1Gm2uju3AsbN93hfex1nEAAuJb5pDBDN2QVCcJ
capitalization: 540981874784286782
stats: BankHashStats {
    num_updated_accounts: 5760,
    num_removed_accounts: 30,
    num_lamports_stored: 1911297845124299,
    total_data_len: 40365454,
    num_executable_accounts: 0
}
```
And the incorrect hash from a `tip-of-master node`:
```
bank frozen: 178114337
hash: 3ncx1EbLUoGPQUoNgwpqYJ8cmaK3BiUwwzN7tLJJ5rmQ
accounts_delta: nrP5aWWJPrrxDnTTavNgXmQBmJf9jK9JDXwiLracJge
signature_count: 2261
last_blockhash: CbE1VP1Gm2uju3AsbN93hfex1nEAAuJb5pDBDN2QVCcJ
capitalization: 540981874784286782,
stats: BankHashStats {
    num_updated_accounts: 5762,
    num_removed_accounts: 31,
    num_lamports_stored: 1911299818513979,
    total_data_len: 41215652,
    num_executable_accounts: 1
}
```
 
We immediately see some differences in the number of updated/removed/executable accounts. These are of course coupled with a difference in data_len, num_lamports_stored and the overall accounts_delta hash.

Adding some extra logging, I determined that the following account state differed between runs:
```
ShadowQL1Pe1bK4XwgUai8HbqW2ngRKmhmV2a8KNYkv
```
And the following three accounts were updated in v1.15/master, but not v1.14:
```
28wAw253EA4y72twSiNXq1jqwdc2aRBZDGiXhnsoZyb4
5Z8EAQYfyUqc9rJdY6Xc78gtCbVbUshygzZgpEag9pqu
FeminazisisUz1rFt4jgTBzdhXAb19VVuo6MjxwdPkiB
```
From these accounts, we can deduce that `5pfUQvvuBv2rEqeJ57f1LDV8JkM8jZjGWTQf4SrjegeP2RJov6iUtg3L3buuM9GQ2YEK8JuvRagVCtJrFKwUrTpU` is the varying transaction:
https://explorer.solana.com/tx/5pfUQvvuBv2rEqeJ57f1LDV8JkM8jZjGWTQf4SrjegeP2RJov6iUtg3L3buuM9GQ2YEK8JuvRagVCtJrFKwUrTpU

Looking at logs from the above explorer link, I see the following for the second instruction:
```
> Program invoked: System Program
  > Program returned success
> ELF error: Incompatible ELF: wrong machine
> Program returned error: "invalid account data for instruction"
```
With v1.14 `solana-leder-tool verify`, I see the same in logs:
```
[2023-02-16T22:27:27.427852134Z DEBUG solana_runtime::message_processor::stable_log] Program BPFLoaderUpgradeab1e11111111111111111111111 invoke [1]
[2023-02-16T22:27:27.427887812Z DEBUG solana_runtime::message_processor::stable_log] Program 11111111111111111111111111111111 invoke [2]
[2023-02-16T22:27:27.428034391Z DEBUG solana_runtime::message_processor::stable_log] Program 11111111111111111111111111111111 success
[2023-02-16T22:27:27.428114764Z DEBUG solana_runtime::message_processor::stable_log] ELF error: Incompatible ELF: wrong machine
[2023-02-16T22:27:27.428119393Z DEBUG solana_runtime::message_processor::stable_log] Program BPFLoaderUpgradeab1e11111111111111111111111 failed: invalid account data for instruction
```
However, I do not see these statements in output from running master:
```
[2023-02-17T19:25:18.053045008Z DEBUG solana_runtime::message_processor::stable_log] Program BPFLoaderUpgradeab1e11111111111111111111111 invoke [1]
[2023-02-17T19:25:18.058619579Z DEBUG solana_runtime::message_processor::stable_log] Program BPFLoaderUpgradeab1e11111111111111111111111 success
```

Thus, it appears that v1.15/master allowed the program upgrade while v1.14 rejected it.

**Comment by steviez** [comment]
@alessandrod @Lichtso - Can one/both of you take a look at the above? I'll poke around a bit, but this looks like something that might be in your realm of expertise.

**Comment by steviez** [comment]
Also, mvines pointed out that another divergence was reported shortly after in slot 178120381:
```
thread 'solReplayStage' panicked at 'We have tried to repair duplicate slot: 178120381 more than 10 times
and are unable to freeze a block with bankhash AM3kMMxQwWhiERhKL9Z6i2Smu6gsCCL9mU8jJc6SsZzt,
instead we have a block with bankhash Some(7o9XK7Sq4TGFkwHqZVprfyaGPapqvWG8LqutzY6kRyVP).

This is most likely a bug in the runtime. At this point manual intervention is needed to make progress. Exiting', core/src/replay_stage.rs:1304:25
```
Checking explorer, it looks like a similar transaction was attempted in that slot:
https://explorer.solana.com/tx/TkddBZgHbFrF55mBRPnXmJ7usDqWmsV5f8ZytnJTAWgXbRP2aWHvenbmRpyEdLEBCJ6ppcXBrAohibfyLQMNj7T

I see the same logs in explorer:
```
> Program invoked: System Program
  > Program returned success
> ELF error: Incompatible ELF: wrong machine
> Program returned error: "invalid account data for instruction"
```
so I feel pretty confident that this second failure was likely the same issue as the first.

**Comment by Lichtso** [comment]
The error is emitted in RBPF here:
https://github.com/solana-labs/rbpf/blob/d0e3b7ef7f6be59c3cb7d0e70e50780281a9e161/src/elf.rs#L613

Lots of things happened during the development of v1.15:
- Last year we developed SBFv2 which was split into (I think 4) different RBPF config flags which were meant to be feature gated later.
- The new ELF parser was also developed around that time and got a config flag.
- We found the executor cache issue and thus never added the features in the monorepo.
- Additionally the config of the RBPF VM was refactored to become immutable so that it could be shared by all programs of a loader to make it more robust and easier to control.
- In the end the releases of RBPF were not properly coordinated with the cut off in the mono repo.

So, I guess this is most likely caused by incorrect feature gating logic around the combination of the new ELF parser and SBFv2.
